<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ziang Zhang" />

<meta name="date" content="2025-02-21" />

<title>Nonlinear Dynamic eQTL analysis on iPSC</title>

<script src="site_libs/header-attrs-2.30/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">fashr-paper</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Nonlinear Dynamic eQTL analysis on
iPSC</h1>
<h4 class="author">Ziang Zhang</h4>
<h4 class="date">2025-02-21</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-12-16
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>fashr-paper-stephenslab/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.2). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20251109code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20251109)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20251109code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20251109)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabfashrpapertree8914a446b959637e6967c5a41e971dd28e43d98ftargetblank8914a44a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/fashr-paper/tree/8914a446b959637e6967c5a41e971dd28e43d98f" target="_blank">8914a44</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabfashrpapertree8914a446b959637e6967c5a41e971dd28e43d98ftargetblank8914a44a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/fashr-paper/tree/8914a446b959637e6967c5a41e971dd28e43d98f" target="_blank">8914a44</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/.DS_Store
    Ignored:    data/toy_example/
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  data/appendixB/
    Untracked:  data/dynamic_eQTL_real/
    Untracked:  output/appendixB/
    Untracked:  output/dynamic_eQTL_real/
    Untracked:  output/toy_example/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/nonlinear_dynamic_eQTL_real.rmd</code>) and HTML
(<code>docs/nonlinear_dynamic_eQTL_real.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/fashr-paper/blob/8914a446b959637e6967c5a41e971dd28e43d98f/analysis/nonlinear_dynamic_eQTL_real.rmd" target="_blank">8914a44</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2025-12-16
</td>
<td>
workflowr::wflow_publish("analysis/nonlinear_dynamic_eQTL_real.rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/fashr-paper/1c558a195c3a6f8d9d0db2314d49bfbb88ccc09a/docs/nonlinear_dynamic_eQTL_real.html" target="_blank">1c558a1</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2025-12-16
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/fashr-paper/blob/a7fabd490cf594553d1965095032acaf042b68a2/analysis/nonlinear_dynamic_eQTL_real.rmd" target="_blank">a7fabd4</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2025-12-16
</td>
<td>
workflowr::wflow_publish("analysis/nonlinear_dynamic_eQTL_real.rmd")
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/fashr-paper/e9f6ed7748c1427db83227d85d276c9e3df12f5d/docs/nonlinear_dynamic_eQTL_real.html" target="_blank">e9f6ed7</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2025-11-09
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/fashr-paper/blob/bb17d9a87a812feb577d7d19fbb6766c15779374/analysis/nonlinear_dynamic_eQTL_real.rmd" target="_blank">bb17d9a</a>
</td>
<td>
Ziang Zhang
</td>
<td>
2025-11-09
</td>
<td>
workflowr::wflow_publish("analysis/nonlinear_dynamic_eQTL_real.rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>knitr::opts_chunk$set(fig.width = 8, fig.height = 6)
library(fashr)
library(biomaRt)
mart &lt;- useEnsembl(biomart = &quot;genes&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)
result_dir &lt;- paste0(getwd(), &quot;/output/dynamic_eQTL_real&quot;)
data_dir &lt;- paste0(getwd(), &quot;/data/dynamic_eQTL_real&quot;)
code_dir &lt;- paste0(getwd(), &quot;/code/dynamic_eQTL_real&quot;)
log_prec &lt;- seq(0,10, by = 0.2)
fine_grid &lt;- sort(c(0, exp(-0.5*log_prec)))</code></pre>
<div id="obtain-the-effect-size-of-eqtls" class="section level2">
<h2>Obtain the effect size of eQTLs</h2>
<p>We use the processed (expression &amp; genotype) data of <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6623972/">Strober
et.al, 2019</a> to perform the eQTL analysis.</p>
<p>For the association testing, we use a linear regression model for
each gene-variant pair at each time point. Following the practice in
Strober et.al, we adjust for the first five PCs.</p>
<p>The code to perform this step can be found in the script
<code>dynamic_eQTL_real/00_eQTLs.R</code> from the code directory.</p>
<p>After this step, we have the effect size of eQTLs for each
gene-variant pair at each time point, as well as its standard error.</p>
</div>
<div id="fitting-fash" class="section level2">
<h2>Fitting FASH</h2>
<p>To fit the FASH model on <span class="math inline">\(\{\beta_i(t_j),
s_{ij}\}_{i\in N,j \in [16]}\)</span>, we consider fitting two FASH
models:</p>
<ul>
<li><p>A FASH model based on first order IWP (testing for dynamic eQTLs:
<span class="math inline">\(H_0: \beta_i(t)=c\)</span>).</p></li>
<li><p>A FASH model based on second order IWP (testing for
nonlinear-dynamic eQTLs: <span class="math inline">\(H_0:
\beta_i(t)=c_1+c_2t\)</span>).</p></li>
</ul>
<p>The code to perform this step can be found in the script
<code>dynamic_eQTL_real/01_fash.R</code> from the code directory.</p>
<p>We will directly load the fitted FASH models from the output
directory.</p>
<pre class="r"><code>load(paste0(result_dir, &quot;/fash_fit2_all.RData&quot;))</code></pre>
<p>We will load the datasets from the fitted FASH object:</p>
<pre class="r"><code>datasets &lt;- fash_fit2$fash_data$data_list
for (i in 1:length(datasets)) {
  datasets[[i]]$SE &lt;- fash_fit2$fash_data$S[[i]]
}
all_genes &lt;- unique(sapply(strsplit(names(datasets), &quot;_&quot;), &quot;[[&quot;, 1))

full_map &lt;- getBM(
  attributes = c(&quot;hgnc_symbol&quot;, &quot;ensembl_gene_id&quot;),
  filters    = &quot;ensembl_gene_id&quot;,
  values     = all_genes,
  mart       = mart
)</code></pre>
<p>In this analysis, we will focus on the <strong>FASH(2)</strong> model
that assumes a <strong>second order IWP</strong> and tests for
<strong>nonlinear dynamic eQTLs</strong>.</p>
<p>Let’s take a quick overview of the fitted FASH model:</p>
<pre class="r"><code>log_prec &lt;- seq(0,10, by = 0.2)
fine_grid &lt;- sort(c(0, exp(-0.5*log_prec)))

fash_fit2 &lt;- fash(Y = &quot;beta&quot;, smooth_var = &quot;time&quot;, S = &quot;SE&quot;, data_list = datasets,
                  num_basis = 20, order = 2, betaprec = 0,
                  pred_step = 1, penalty = 10, grid = fine_grid,
                  num_cores = num_cores, verbose = TRUE)
save(fash_fit2, file = &quot;./results/fash_fit2_all.RData&quot;)</code></pre>
<pre class="r"><code>fash_fit2</code></pre>
<pre><code>Fitted fash Object
-------------------
Number of datasets: 1009173
Likelihood: gaussian
Number of PSD grid values: 52 (initial), 9 (non-trivial)
Order of Integrated Wiener Process (IWP): 2</code></pre>
<p>As well as the estimated priors:</p>
<pre class="r"><code>fash_fit2$prior_weights</code></pre>
<pre><code>          psd prior_weight
1 0.000000000 9.872378e-01
2 0.006737947 1.209985e-04
3 0.020241911 5.257484e-03
4 0.022370772 2.211188e-03
5 0.055023220 1.113735e-03
6 0.060810063 3.628552e-03
7 0.246596964 3.827515e-04
8 0.449328964 1.965219e-05
9 1.000000000 2.786579e-05</code></pre>
</div>
<div id="problem-with-pi_0-estimation" class="section level2">
<h2>Problem with <span class="math inline">\(\pi_0\)</span>
estimation</h2>
<p>The original MLE estimated <span class="math inline">\(\pi_0\)</span>
is 0.9872378. This could be under-estimated due to
model-misspecification under the alternative hypothesis. To account for
this, we will consider the following approaches:</p>
<p>(i): A conservative estimate of <span
class="math inline">\(\pi_0\)</span> based on the BF procedure:</p>
<pre class="r"><code>fash_fit2_update &lt;- BF_update(fash_fit2, plot = FALSE)
fash_fit2_update$prior_weights
save(fash_fit2_update, file = paste0(result_dir, &quot;/fash_fit2_update.RData&quot;))</code></pre>
<p>The conservative estimate is 0.9992301, which is much more
conservative.</p>
<p>(ii): Instead of looking at the FDR which is based on the estimated
<span class="math inline">\(\pi_0\)</span>, we can use the minimum local
false sign rate (<span class="math inline">\(\text{min-lfsr}_i\)</span>)
to measure significance: <span class="math display">\[
\text{min-lfsr}_i = \min_{t} \left\{ \text{lfsr}(W_i(t)) \right\},
\]</span> where <span class="math inline">\(W_i(t) = \beta_i(t) -
\beta_i(0)\)</span>.</p>
</div>
<div id="detecting-nonlinear-dynamic-eqtls" class="section level2">
<h2>Detecting Nonlinear dynamic eQTLs</h2>
<p>We will use the updated FASH model (2) to detect nonlinear dynamic
eQTLs.</p>
<pre class="r"><code>alpha &lt;- 0.05
test2 &lt;- fdr_control(fash_fit2_update, alpha = alpha, plot = F)</code></pre>
<pre><code>96 datasets are significant at alpha level 0.05. Total datasets tested: 1009173. </code></pre>
<pre class="r"><code>fash_highlighted2 &lt;- test2$fdr_results$index[test2$fdr_results$FDR &lt;= alpha]

test2_before &lt;- fdr_control(fash_fit2, alpha = alpha, plot = F)</code></pre>
<pre><code>343 datasets are significant at alpha level 0.05. Total datasets tested: 1009173. </code></pre>
<pre class="r"><code>fash_highlighted2_before &lt;- test2_before$fdr_results$index[test2_before$fdr_results$FDR &lt;= alpha]</code></pre>
<p>How many pairs are detected?</p>
<pre class="r"><code>pairs_highlighted2 &lt;- names(datasets)[fash_highlighted2]
length(pairs_highlighted2)</code></pre>
<pre><code>[1] 96</code></pre>
<pre class="r"><code>length(pairs_highlighted2)/length(datasets)</code></pre>
<pre><code>[1] 9.51274e-05</code></pre>
<p>What is the number before the BF adjustment?</p>
<pre class="r"><code>pairs_highlighted2_before &lt;- names(datasets)[fash_highlighted2_before]
length(pairs_highlighted2_before)</code></pre>
<pre><code>[1] 343</code></pre>
<pre class="r"><code>length(pairs_highlighted2_before)/length(datasets)</code></pre>
<pre><code>[1] 0.0003398823</code></pre>
<p>How many unique genes are detected?</p>
<pre class="r"><code>genes_highlighted2 &lt;- unique(sapply(strsplit(pairs_highlighted2, &quot;_&quot;), &quot;[[&quot;, 1))
length(genes_highlighted2)</code></pre>
<pre><code>[1] 27</code></pre>
<pre class="r"><code>length(genes_highlighted2)/length(all_genes)</code></pre>
<pre><code>[1] 0.004243948</code></pre>
<p>Before the BF adjustment?</p>
<pre class="r"><code>genes_highlighted2_before &lt;- unique(sapply(strsplit(pairs_highlighted2_before, &quot;_&quot;), &quot;[[&quot;, 1))
length(genes_highlighted2_before)</code></pre>
<pre><code>[1] 63</code></pre>
<pre class="r"><code>length(genes_highlighted2_before)/length(all_genes)</code></pre>
<pre><code>[1] 0.009902546</code></pre>
<p>Visualize top-ranked pairs:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/nonlinear_dynamic_examples-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Some examples of null pairs:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/nonlinear_null_examples-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="comparing-with-strober-et.al" class="section level2">
<h2>Comparing with Strober et.al</h2>
<p>We will compare the detected dynamic eQTLs with the results from
Strober et.al.</p>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.1     ✔ stringr   1.5.2
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
✖ dplyr::select() masks biomaRt::select()
✖ tidyr::unite()  masks ggVennDiagram::unite()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<p>Let’s take a look at the overlap between the two methods used in
Strober et.al and FASH (order 2):</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/venn_diagram-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Let’s take a look at the 4 pairs that are least significant from
FASH:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/only_fash_least-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Let’s also look at the genes that were missed by FASH, but detected
by Strober et.al. In this case, we will pick the most significant pair
for each gene in FASH:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/miss_fash_most-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="classifying-nonlinear-dynamic-eqtls" class="section level2">
<h2>Classifying nonlinear dynamic eQTLs</h2>
<p>Following the definition in Strober et.al, we will classify the
detected eQTLs into different categories:</p>
<ul>
<li><p><em>Early</em>: eQTLs with strongest effect during the first
three days: <span class="math inline">\(\max_{t\leq3} |\beta(t)| -
\max_{t&gt; 3} |\beta(t)| &gt; 0\)</span>.</p></li>
<li><p><em>Late</em>: eQTLs with strongest effect during the last four
days: <span class="math inline">\(\max_{t\geq 12} |\beta(t)| -
\max_{t&lt; 12} |\beta(t)| &gt; 0\)</span>.</p></li>
<li><p><em>Middle</em>: eQTLs with strongest effect during days 4-11:
<span class="math inline">\(\max_{4\leq t\leq 11} |\beta(t)| -
\max_{t&gt; 11 | t&lt; 4} |\beta(t)| &gt; 0\)</span>.</p></li>
<li><p><em>Switch</em>: eQTLs with effect sign switch during the time
course: ${(t)<sup>+,(t)</sup>-}-c $ where <span
class="math inline">\(c\)</span> is a threshold that we set to 0.25
(which means with two alleles, the maximal difference of effect size is
at least <span class="math inline">\(\geq
2\times\min\{\max\beta(t)^+,\max\beta(t)^-\}\times2 \geq  2 \times
0.25  \times 2 = 1\)</span>).</p></li>
</ul>
<p>We first take a look at the significant pairs detected by FASH (order
2), and classify them based on the false sign rate (lfsr):</p>
<pre class="r"><code>smooth_var_refined = seq(0,15, by = 0.1)
functional_early &lt;- function(x){
  max(abs(x[smooth_var_refined &lt;= 3])) - max(abs(x[smooth_var_refined &gt; 3]))
}
testing_early_nonlin_dyn &lt;- testing_functional(functional_early,
                                              lfsr_cal = function(x){mean(x &lt;= 0)},
                                              fash = fash_fit2,
                                              indices = fash_highlighted2,
                                              smooth_var = smooth_var_refined)</code></pre>
<p>How many pairs and how many unique genes are classified as early
dynamic eQTLs?</p>
<pre class="r"><code>load(paste0(result_dir, &quot;/classify_nonlin_dyn_eQTLs_early.RData&quot;))
early_indices &lt;- testing_early_nonlin_dyn$indices[testing_early_nonlin_dyn$cfsr &lt;= alpha]
length(early_indices)</code></pre>
<pre><code>[1] 0</code></pre>
<pre class="r"><code>early_genes &lt;- unique(sapply(strsplit(names(datasets)[early_indices], &quot;_&quot;), &quot;[[&quot;, 1))
length(early_genes)</code></pre>
<pre><code>[1] 0</code></pre>
<p>How many pairs are classified as middle dynamic eQTLs?</p>
<pre class="r"><code>functional_middle &lt;- function(x){
  max(abs(x[smooth_var_refined &lt;= 11 &amp; smooth_var_refined &gt;= 4])) - max(abs(x[smooth_var_refined &gt; 11]), abs(x[smooth_var_refined &lt; 4]))
}
testing_middle_nonlin_dyn &lt;- testing_functional(functional_middle, 
                                               lfsr_cal = function(x){mean(x &lt;= 0)},
                                               fash = fash_fit2, 
                                               indices = fash_highlighted2, 
                                               num_cores = num_cores,
                                               smooth_var = smooth_var_refined)</code></pre>
<pre class="r"><code>load(paste0(result_dir, &quot;/classify_nonlin_dyn_eQTLs_middle.RData&quot;))
middle_indices &lt;- testing_middle_nonlin_dyn$indices[testing_middle_nonlin_dyn$cfsr &lt;= alpha]
length(middle_indices)</code></pre>
<pre><code>[1] 32</code></pre>
<pre class="r"><code>middle_genes &lt;- unique(sapply(strsplit(names(datasets)[middle_indices], &quot;_&quot;), &quot;[[&quot;, 1))
length(middle_genes)</code></pre>
<pre><code>[1] 7</code></pre>
<p>Take a look at their results:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/middle_examples-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>How many pairs are classified as late dynamic eQTLs?</p>
<pre class="r"><code>functional_late &lt;- function(x){
  max(abs(x[smooth_var_refined &gt;= 12])) - max(abs(x[smooth_var_refined &lt; 12]))
}
testing_late_nonlin_dyn &lt;- testing_functional(functional_late, 
                                             lfsr_cal = function(x){mean(x &lt;= 0)},
                                             fash = fash_fit2, 
                                             indices = fash_highlighted2, 
                                             num_cores = num_cores,
                                             smooth_var = smooth_var_refined)</code></pre>
<pre class="r"><code>load(paste0(result_dir, &quot;/classify_nonlin_dyn_eQTLs_late.RData&quot;))
late_indices &lt;- testing_late_nonlin_dyn$indices[testing_late_nonlin_dyn$cfsr &lt;= alpha]
length(late_indices)</code></pre>
<pre><code>[1] 45</code></pre>
<pre class="r"><code>late_genes &lt;- unique(sapply(strsplit(names(datasets)[late_indices], &quot;_&quot;), &quot;[[&quot;, 1))
length(late_genes)</code></pre>
<pre><code>[1] 14</code></pre>
<p>Let’s take a look at the top-ranked late dynamic eQTLs:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/late_examples-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>How many pairs and how many unique genes are classified as switch
dynamic eQTLs?</p>
<pre class="r"><code>switch_threshold &lt;- 0.25
functional_switch &lt;- function(x){
  # compute the radius of x, measured by deviation from 0 from below and from above
  x_pos &lt;- x[x &gt; 0]
  x_neg &lt;- x[x &lt; 0]
  if(length(x_pos) == 0 || length(x_neg) == 0){
    return(0)
  }
  min(max(abs(x_pos)), max(abs(x_neg))) - switch_threshold
}
testing_switch_nonlin_dyn &lt;- testing_functional(functional_switch, 
                                               lfsr_cal = function(x){mean(x &lt;= 0)},
                                               fash = fash_fit2, 
                                               indices = fash_highlighted2, 
                                               num_cores = num_cores,
                                               smooth_var = smooth_var_refined)</code></pre>
<pre class="r"><code>load(paste0(result_dir, &quot;/classify_nonlin_dyn_eQTLs_switch.RData&quot;))
switch_indices &lt;- testing_switch_nonlin_dyn$indices[testing_switch_nonlin_dyn$cfsr &lt;= alpha]
length(switch_indices)</code></pre>
<pre><code>[1] 9</code></pre>
<pre class="r"><code>switch_genes &lt;- unique(sapply(strsplit(names(datasets)[switch_indices], &quot;_&quot;), &quot;[[&quot;, 1))
length(switch_genes)</code></pre>
<pre><code>[1] 9</code></pre>
<p>Let’s take a look at the top-ranked switch dynamic eQTLs:</p>
<p><img src="figure/nonlinear_dynamic_eQTL_real.rmd/switch_examples-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="gene-set-enrichment-analysis" class="section level2">
<h2>Gene Set Enrichment Analysis</h2>
<pre class="r"><code>library(clusterProfiler)
library(tidyverse)
library(msigdbr)
library(org.Hs.eg.db)  # Assuming human genes
library(biomaRt)
library(cowplot)
# Retrieve Hallmark gene sets for Homo sapiens
m_t2g &lt;- msigdbr(species = &quot;Homo sapiens&quot;, category = &quot;H&quot;) %&gt;% 
  dplyr::select(gs_name, entrez_gene)
mart &lt;- useMart(&quot;ensembl&quot;, dataset = &quot;hsapiens_gene_ensembl&quot;)

## A function to check gene-enrichment
## Retrieve Hallmark gene sets for Homo sapiens (use Ensembl IDs)
m_t2g &lt;- msigdbr(species = &quot;Homo sapiens&quot;, category = &quot;H&quot;) %&gt;% 
  mutate(ensembl_use = dplyr::coalesce(ensembl_gene, db_ensembl_gene)) %&gt;% 
  dplyr::filter(!is.na(ensembl_use)) %&gt;% 
  dplyr::select(gs_name, ensembl_use) %&gt;%
  dplyr::distinct()    # one row per (pathway, Ensembl) pair


## A function to check gene-enrichment using Ensembl IDs,
## forcing the universe to be exactly background_gene
enrich_set &lt;- function(genes_selected,
                       background_gene,
                       q_val_cutoff = 0.05,
                       pvalueCutoff = 0.05) {

  # ensure character vectors &amp; unique
  genes_selected_raw  &lt;- unique(as.character(genes_selected))
  background_gene_raw &lt;- unique(as.character(background_gene))

  # we&#39;ll enforce that the test universe is exactly these:
  universe_for_test &lt;- background_gene_raw

  # genes that are in Hallmark already
  hallmark_genes &lt;- unique(m_t2g$ensembl_use)

  # background genes that are NOT in Hallmark (no annotation)
  bg_not_in_hallmark &lt;- setdiff(universe_for_test, hallmark_genes)

  # extend TERM2GENE so that *all* background genes appear in it at least once
  # add them to a dummy pathway that we will later drop
  dummy_id &lt;- &quot;__DUMMY_BACKGROUND__&quot;

  if (length(bg_not_in_hallmark) &gt; 0) {
    dummy_t2g &lt;- tibble(
      gs_name     = dummy_id,
      ensembl_use = bg_not_in_hallmark
    )
    TERM2GENE_full &lt;- bind_rows(m_t2g, dummy_t2g)
  } else {
    TERM2GENE_full &lt;- m_t2g
  }

  # also make sure selected genes are a subset of the universe
  genes_sel_used &lt;- intersect(genes_selected_raw, universe_for_test)

  # message(&quot;Original selected genes: &quot;, length(genes_selected_raw),
  #         &quot; ; used in enrichment: &quot;, length(genes_sel_used))
  # message(&quot;Universe size (background_gene): &quot;, length(universe_for_test))

  enrich_res &lt;- enricher(
    gene          = genes_sel_used,
    TERM2GENE     = TERM2GENE_full,
    universe      = universe_for_test,
    pAdjustMethod = &quot;BH&quot;,
    qvalueCutoff  = q_val_cutoff,
    pvalueCutoff  = pvalueCutoff
  )

  if (is.null(enrich_res) || nrow(enrich_res@result) == 0L) {
    return(enrich_res)
  }

  df &lt;- enrich_res@result

  # drop the dummy term from results
  df &lt;- df %&gt;% dplyr::filter(ID != dummy_id)

  # keep original ratios for reference
  df$GeneRatio_orig &lt;- df$GeneRatio
  df$BgRatio_orig   &lt;- df$BgRatio

  # now recompute ratios using exactly:
  #   - denominator for GeneRatio = number of selected genes in universe
  #   - denominator for BgRatio   = number of background genes
  n_sel_total &lt;- length(genes_sel_used)
  n_bg_total  &lt;- length(universe_for_test)

  df$GeneRatio_fixed &lt;- paste0(df$Count,   &quot;/&quot;, n_sel_total)
  df$BgRatio_fixed   &lt;- paste0(df$setSize, &quot;/&quot;, n_bg_total)

  enrich_res@result &lt;- df
  enrich_res
}</code></pre>
<p>Among all the genes highlighted by FASH:</p>
<pre class="r"><code>result &lt;- enrich_set(genes_selected = genes_highlighted2, background_gene = all_genes)
result@result %&gt;% 
  filter(pvalue &lt; 0.05) %&gt;%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)</code></pre>
<pre><code>                                GeneRatio BgRatio     pvalue    qvalue
HALLMARK_MYOGENESIS                  2/27 67/6362 0.03237802 0.1841743
HALLMARK_FATTY_ACID_METABOLISM       2/27 72/6362 0.03694955 0.1841743
HALLMARK_ESTROGEN_RESPONSE_LATE      2/27 79/6362 0.04374140 0.1841743</code></pre>
<p>Among the genes highlighted by FASH that are classified as middle
dynamic eQTLs:</p>
<pre class="r"><code>result &lt;- enrich_set(genes_selected = middle_genes, background_gene = all_genes)
result@result %&gt;% 
  filter(pvalue &lt; 0.05) %&gt;%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)</code></pre>
<pre><code>                                   GeneRatio BgRatio     pvalue     qvalue
HALLMARK_INTERFERON_ALPHA_RESPONSE       1/7 27/6362 0.02934573 0.07629487
HALLMARK_ALLOGRAFT_REJECTION             1/7 36/6362 0.03896214 0.07629487
HALLMARK_INFLAMMATORY_RESPONSE           1/7 43/6362 0.04638500 0.07629487</code></pre>
<p>Among the genes highlighted by FASH that are classified as late
dynamic eQTLs:</p>
<pre class="r"><code>result &lt;- enrich_set(genes_selected = late_genes, background_gene = all_genes)
result@result %&gt;% 
  filter(pvalue &lt; 0.05) %&gt;%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)</code></pre>
<pre><code>                                 GeneRatio BgRatio      pvalue     qvalue
HALLMARK_MYOGENESIS                   2/14 67/6362 0.009163424 0.04822855
HALLMARK_IL6_JAK_STAT3_SIGNALING      1/14 21/6362 0.045278647 0.10516379</code></pre>
<p>Among the genes highlighted by FASH that are classified as switch
dynamic eQTLs:</p>
<pre class="r"><code>result &lt;- enrich_set(genes_selected = switch_genes, background_gene = all_genes)
result@result %&gt;% 
  filter(pvalue &lt; 0.05) %&gt;%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)</code></pre>
<pre><code>                                   GeneRatio BgRatio      pvalue     qvalue
HALLMARK_MYOGENESIS                      2/9 67/6362 0.003750287 0.02763370
HALLMARK_IL6_JAK_STAT3_SIGNALING         1/9 21/6362 0.029336610 0.08333513
HALLMARK_INTERFERON_ALPHA_RESPONSE       1/9 27/6362 0.037576748 0.08333513
HALLMARK_ALLOGRAFT_REJECTION             1/9 36/6362 0.049820388 0.08333513</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.6.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] cowplot_1.2.0          org.Hs.eg.db_3.21.0    AnnotationDbi_1.70.0  
 [4] IRanges_2.42.0         S4Vectors_0.46.0       Biobase_2.68.0        
 [7] BiocGenerics_0.54.1    generics_0.1.4         msigdbr_25.1.1        
[10] clusterProfiler_4.16.0 lubridate_1.9.4        forcats_1.0.1         
[13] stringr_1.5.2          dplyr_1.1.4            purrr_1.1.0           
[16] readr_2.1.5            tidyr_1.3.1            tibble_3.3.0          
[19] ggplot2_4.0.0          tidyverse_2.0.0        ggVennDiagram_1.5.4   
[22] biomaRt_2.64.0         fashr_0.1.42           workflowr_1.7.2       

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      rstudioapi_0.17.1       jsonlite_2.0.0         
  [4] magrittr_2.0.4          ggtangle_0.0.8          farver_2.1.2           
  [7] rmarkdown_2.30          fs_1.6.6                vctrs_0.6.5            
 [10] memoise_2.0.1           ggtree_3.16.3           mixsqp_0.3-54          
 [13] htmltools_0.5.8.1       progress_1.2.3          curl_7.0.0             
 [16] gridGraphics_0.5-1      sass_0.4.10             bslib_0.9.0            
 [19] plyr_1.8.9              httr2_1.2.1             cachem_1.1.0           
 [22] TMB_1.9.18              igraph_2.2.0            whisker_0.4.1          
 [25] lifecycle_1.0.4         pkgconfig_2.0.3         gson_0.1.0             
 [28] Matrix_1.7-3            R6_2.6.1                fastmap_1.2.0          
 [31] GenomeInfoDbData_1.2.14 digest_0.6.37           numDeriv_2016.8-1.1    
 [34] aplot_0.2.9             enrichplot_1.28.4       colorspace_2.1-2       
 [37] patchwork_1.3.2         ps_1.9.1                rprojroot_2.1.1        
 [40] irlba_2.3.5.1           RSQLite_2.4.3           filelock_1.0.3         
 [43] labeling_0.4.3          timechange_0.3.0        httr_1.4.7             
 [46] compiler_4.5.1          bit64_4.6.0-1           withr_3.0.2            
 [49] S7_0.2.0                BiocParallel_1.42.2     DBI_1.2.3              
 [52] R.utils_2.13.0          rappdirs_0.3.3          tools_4.5.1            
 [55] ape_5.8-1               httpuv_1.6.16           R.oo_1.27.1            
 [58] glue_1.8.0              callr_3.7.6             nlme_3.1-168           
 [61] GOSemSim_2.34.0         promises_1.3.3          grid_4.5.1             
 [64] getPass_0.2-4           reshape2_1.4.4          fgsea_1.34.2           
 [67] gtable_0.3.6            tzdb_0.5.0              R.methodsS3_1.8.2      
 [70] data.table_1.17.8       hms_1.1.3               xml2_1.4.0             
 [73] XVector_0.48.0          ggrepel_0.9.6           pillar_1.11.1          
 [76] babelgene_22.9          yulab.utils_0.2.1       later_1.4.4            
 [79] splines_4.5.1           treeio_1.32.0           BiocFileCache_2.16.2   
 [82] lattice_0.22-7          bit_4.6.0               tidyselect_1.2.1       
 [85] GO.db_3.21.0            Biostrings_2.76.0       knitr_1.50             
 [88] git2r_0.36.2            xfun_0.54               LaplacesDemon_16.1.6   
 [91] stringi_1.8.7           UCSC.utils_1.4.0        lazyeval_0.2.2         
 [94] ggfun_0.2.0             yaml_2.3.10             evaluate_1.0.5         
 [97] codetools_0.2-20        qvalue_2.40.0           ggplotify_0.1.3        
[100] cli_3.6.5               processx_3.8.6          jquerylib_0.1.4        
[103] dichromat_2.0-0.1       Rcpp_1.1.0              GenomeInfoDb_1.44.3    
[106] dbplyr_2.5.1            png_0.1-8               parallel_4.5.1         
[109] assertthat_0.2.1        blob_1.2.4              prettyunits_1.2.0      
[112] DOSE_4.2.0              tidytree_0.4.6          scales_1.4.0           
[115] crayon_1.5.3            rlang_1.1.6             fastmatch_1.1-6        
[118] KEGGREST_1.48.1        </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
