---
title: "Dynamic eQTL analysis on iPSC"
author: "Ziang Zhang"
date: "2025-02-21"
output: 
  workflowr::wflow_html:
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(fig.width = 8, fig.height = 6)
library(fashr)
result_dir <- paste0(getwd(), "/output/dynamic_eQTL_real")
data_dir <- paste0(getwd(), "/data/dynamic_eQTL_real")
code_dir <- paste0(getwd(), "/code/dynamic_eQTL_real")
log_prec <- seq(0,10, by = 0.2)
fine_grid <- sort(c(0, exp(-0.5*log_prec)))
```

## Obtain the effect size of eQTLs

We use the processed (expression \& genotype) data of [Strober et.al, 2019](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6623972/) to perform the eQTL analysis.

For the association testing, we use a linear regression model for each gene-variant pair at each time point.
Following the practice in Strober et.al, we adjust for the first five PCs.

The code to perform this step can be found in the script `dynamic_eQTL_real/00_eQTLs.R` from the code directory.

After this step, we have the effect size of eQTLs for each gene-variant pair at each time point, as well as its standard error.

## Fitting FASH

To fit the FASH model on $\{\beta_i(t_j), s_{ij}\}_{i\in N,j \in [16]}$, we consider fitting two FASH models:

- A FASH model based on first order IWP (testing for dynamic eQTLs: $H_0: \beta_i(t)=c$).

- A FASH model based on second order IWP (testing for nonlinear-dynamic eQTLs: $H_0: \beta_i(t)=c_1+c_2t$).

The code to perform this step can be found in the script `dynamic_eQTL_real/01_fash.R` from the code directory.

We will directly load the fitted FASH models from the output directory.

```{r}
load(paste0(result_dir, "/fash_fit1_all.RData"))
```

We will load the datasets from the fitted FASH object:

```{r}
datasets <- fash_fit1$fash_data$data_list
for (i in 1:length(datasets)) {
  datasets[[i]]$SE <- fash_fit1$fash_data$S[[i]]
}
all_genes <- unique(sapply(strsplit(names(datasets), "_"), "[[", 1))
```


In this analysis, we will focus on the **FASH(1)** model that assumes a **first order IWP** and **tests for dynamic eQTLs**.

Let's take a quick overview of the fitted FASH model:

```{r, eval=FALSE}
log_prec <- seq(0,10, by = 0.2)
fine_grid <- sort(c(0, exp(-0.5*log_prec)))

fash_fit1 <- fash(Y = "beta", smooth_var = "time", S = "SE", data_list = datasets,
                  num_basis = 20, order = 1, betaprec = 0,
                  pred_step = 1, penalty = 10, grid = fine_grid,
                  num_cores = num_cores, verbose = TRUE)
save(fash_fit1, file = "./results/fash_fit1_all.RData")
```


```{r}
fash_fit1
```

As well as the estimated priors:
```{r}
fash_fit1$prior_weights
```


## Problem with $\pi_0$ estimation

If we measure the significance using the false discovery rate, then it is sensitive to the value of $\pi_0$.
The estimated $\pi_0$ is `r fash_fit1$prior_weights$prior_weight[1]`, which is way too small to be realistic.

One likely reason could be due to model-misspecification under the alternative hypothesis.
To account for this, we will consider the following approaches:

(i): Computing a conservative estimate of $\pi_0$ based on the BF procedure:

```{r eval=FALSE}
fash_fit1_update <- BF_update(fash_fit1, plot = FALSE)
fash_fit1_update$prior_weights
save(fash_fit1_update, file = paste0(result_dir, "/fash_fit1_update.RData"))
```

```{r echo=FALSE}
load(paste0(result_dir, "/fash_fit1_update.RData"))
```

The conservative estimate is `r fash_fit1_update$prior_weights$prior_weight[1]`, which is much more realistic.

(ii): Instead of looking at the FDR which is based on the estimated $\pi_0$, we can use the minimum local false sign rate ($\text{min-lfsr}_i$) to measure significance:
$$
\text{min-lfsr}_i = \min_{t} \left\{ \text{lfsr}(W_i(t)) \right\},
$$
where $W_i(t) = \beta_i(t) - \beta_i(0)$.

Let's compute the significance using the minimum local false sign rate ($\text{min-lfsr}_i$):

```{r eval=FALSE}
smooth_var_refined = seq(0,15, by = 0.1)
min_lfsr_summary1 <- min_lfsr_summary(fash_fit1, num_cores = num_cores, smooth_var = smooth_var_refined)
save(min_lfsr_summary1, file = "./results/min_lfsr_summary1.RData")
min_lfsr_summary1_update <- min_lfsr_summary(fash_fit1_update, num_cores = num_cores, smooth_var = smooth_var_refined)
save(min_lfsr_summary1_update, file = "./results/min_lfsr_summary1_update.RData")
```

Let's visualize how the min-lfsr compares with the local false discovery rate (lfdr):

```{r}
load(paste0(result_dir, "/min_lfsr_summary1.RData"))
# sample some indices for easy visualization
sample_indices <- sample(1:length(min_lfsr_summary1$min_lfsr), 1e5)
min_lfsr1_unadj <- min_lfsr_summary1$min_lfsr[sample_indices]
lfdr1_vec_unadj <- fash_fit1$lfdr[min_lfsr_summary1$index][sample_indices]
plot(lfdr1_vec_unadj, min_lfsr1_unadj, 
     pch = 20, cex = 0.1,
     ylim = c(0,1), xlim = c(0,1),
     xlab = "lfdr", ylab = "min-lfsr", main = "lfdr vs min-lfsr (Un-adjusted)")
lines(c(0,1), c(0,1), col = "red")
```

Let's also visualize the min-lfsr and the lfdr from the BF-updated model:

```{r}
load(paste0(result_dir, "/min_lfsr_summary1_update.RData"))
min_lfsr1 <- min_lfsr_summary1_update$min_lfsr[sample_indices]
lfdr1_vec <- fash_fit1_update$lfdr[min_lfsr_summary1_update$index[sample_indices]]
plot(lfdr1_vec, min_lfsr1, 
     pch = 20, cex = 0.1,
     ylim = c(0,1), xlim = c(0,1),
     xlab = "lfdr", ylab = "min-lfsr", main = "lfdr vs min-lfsr (BF-adjusted)")
lines(c(0,1), c(0,1), col = "red")
```


Indeed, the min-lfsr tends to be more conservative than the lfdr, especially when $\hat{\pi_0}$ has not been adjusted using the BF procedure.

Let's also assess how does the BF-update of $\pi_0$ affect the min-lfsr.

```{r}
plot(min_lfsr1_unadj, min_lfsr1, 
     pch = 20, cex = 0.1,
     ylim = c(0,1), xlim = c(0,1),
     xlab = "min-lfsr (unadjusted)", ylab = "min-lfsr", main = "Effect of BF adjustment on min-lfsr")
lines(c(0,1), c(0,1), col = "red")
```



## Detecting dynamic eQTLs

We will use the updated FASH model (1) to detect dynamic eQTLs.

```{r}
alpha <- 0.05
test1 <- fdr_control(fash_fit1_update, alpha = alpha, plot = F)
fash_highlighted1 <- test1$fdr_results$index[test1$fdr_results$FDR <= alpha]
```

How many pairs are detected as dynamic eQTLs?

```{r}
pairs_highlighted1 <- names(datasets)[fash_highlighted1]
length(pairs_highlighted1)
length(pairs_highlighted1)/length(datasets)
```

How many unique genes are detected?

```{r}
genes_highlighted1 <- unique(sapply(strsplit(pairs_highlighted1, "_"), "[[", 1))
length(genes_highlighted1)
length(genes_highlighted1)/length(all_genes)
```


Let's see how many pairs and genes remain significant after controlling the min-lfsr:

```{r}
fash_highlighted1_lfsr <- min_lfsr_summary1_update$index[min_lfsr_summary1_update$fsr <= alpha]
pairs_highlighted1_lfsr <- names(datasets)[fash_highlighted1_lfsr]
length(pairs_highlighted1_lfsr)
length(pairs_highlighted1_lfsr)/length(datasets)
```

```{r}
genes_highlighted1_lfsr <- unique(sapply(strsplit(pairs_highlighted1_lfsr, "_"), "[[", 1))
length(genes_highlighted1_lfsr)
length(genes_highlighted1_lfsr)/length(all_genes)
```

It seems like once $\hat{\pi_0}$ is adjusted, there is not much difference between measuring significance using the min-lfsr or the lfdr.
From now on, we will consider the pairs that are significant using the lfdr.


Visualize top-ranked pairs:

```{r echo=FALSE}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_gene <- (genes_highlighted1)[i] #sample(genes_highlighted1, 1)
  pairs_of_selected_gene <- grep(selected_gene, pairs_highlighted1, value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit1_update$lfdr[selected_indices] == min(fash_fit1_update$lfdr[selected_indices]))][1]
  fitted_result <- predict(fash_fit1_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  abline(h = mean(datasets[[selected_index]]$y), col = "blue", lty = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```


Some examples of null pairs:

```{r}
genes_not_highlighted1 <- setdiff(all_genes, genes_highlighted1)
pairs_not_highlighted1 <- setdiff(names(datasets), pairs_highlighted1)
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_gene <- (genes_not_highlighted1)[i] #sample(genes_highlighted1, 1)
  pairs_of_selected_gene <- grep(selected_gene, pairs_not_highlighted1, value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit1_update$lfdr[selected_indices] == max(fash_fit1_update$lfdr[selected_indices]))][1]
  fitted_result <- predict(fash_fit1_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  abline(h = mean(datasets[[selected_index]]$y), col = "blue", lty = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```



## Comparing with Strober et.al

We will compare the detected dynamic eQTLs with the results from Strober et.al.

```{r echo=FALSE}
library(ggVennDiagram)
library(tidyverse)
strober_nonlinear <- read.delim(paste0(data_dir, "/strober_nonlinear/non_linear_dynamic_eqtls_5_pc.txt"))
strober_nonlinear$key <- paste0(strober_nonlinear$ensamble_id, "_", strober_nonlinear$rs_id)
strober_nonlinear_highlighted <- strober_nonlinear %>% filter(eFDR <= alpha)
pairs_highlighted_strober_nonlinear <- strober_nonlinear_highlighted$key
genes_highlighted_strober_nonlinear <- unique(sapply(strsplit(pairs_highlighted_strober_nonlinear, "_"), "[[", 1))
strober_linear <- read.delim(paste0(data_dir, "/strober_linear/linear_dynamic_eqtls_5_pc.txt"))
strober_linear$key <- paste0(strober_linear$ensamble_id, "_", strober_linear$rs_id)
strober_linear_highlighted <- strober_linear %>% filter(eFDR <= alpha)
pairs_highlighted_strober_linear <- strober_linear_highlighted$key
genes_highlighted_strober_linear <- unique(sapply(strsplit(pairs_highlighted_strober_linear, "_"), "[[", 1))
```

Let's take a look at the overlap between the two methods used in Strober et.al and FASH (order 1):

```{r}
gene_sets <- list(
  "Strober (Nonlinear)" = genes_highlighted_strober_nonlinear,
  "Strober (Linear)" = genes_highlighted_strober_linear,
  "FASH (1)" = genes_highlighted1
)
ggVennDiagram(gene_sets, label = "both") +
  scale_fill_gradient(low="grey90",high = "red") +
  theme(legend.position = "right")  # Move legend to the right
```

There is a large number of genes only detected by FASH (order 1). 
Let's take a look at the 4 pairs that are least significant from FASH:

```{r echo=FALSE}
gene_fashr_only <- setdiff(genes_highlighted1, union(genes_highlighted_strober_nonlinear, genes_highlighted_strober_linear))
## produce four plots for four genes
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_gene <- gene_fashr_only[i]
  pairs_of_selected_gene <- grep(selected_gene, pairs_highlighted1, value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit1_update$lfdr[selected_indices] == max(fash_fit1_update$lfdr[selected_indices]))][1]
  cat(fash_fit1_update$lfdr[selected_index]); cat("\n")
  fitted_result <- predict(fash_fit1_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  inv_weights <- (1/datasets[[selected_index]]$SE)/sum(1/datasets[[selected_index]]$SE)
  abline(h = sum(datasets[[selected_index]]$y*inv_weights), col = "blue", lty = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```

Let's also take a look at the 4 pairs that are most significant from FASH:

```{r echo=FALSE}
gene_fashr_only <- setdiff(genes_highlighted1, union(genes_highlighted_strober_nonlinear, genes_highlighted_strober_linear))
## produce four plots for four genes
par(mfrow = c(2,2))
for (i in 5:8) {
  selected_gene <- gene_fashr_only[i]
  pairs_of_selected_gene <- grep(selected_gene, pairs_highlighted1, value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit1_update$lfdr[selected_indices] == min(fash_fit1_update$lfdr[selected_indices]))][1]
  cat(fash_fit1_update$lfdr[selected_index]); cat("\n")
  fitted_result <- predict(fash_fit1_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  inv_weights <- (1/datasets[[selected_index]]$SE)/sum(1/datasets[[selected_index]]$SE)
  abline(h = sum(datasets[[selected_index]]$y*inv_weights), col = "blue", lty = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}

par(mfrow = c(1,1))
```



Let's also look at the genes that were missed by FASH, but detected by Strober et.al.
In this case, we will pick the most significant pair for each gene in FASH:

```{r echo=FALSE}
gene_missed <- setdiff(union(genes_highlighted_strober_nonlinear, genes_highlighted_strober_linear), genes_highlighted1)
## produce four plots for four genes
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_gene <- gene_missed[i]
  pairs_of_selected_gene <- grep(selected_gene, names(datasets), value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit1_update$lfdr[selected_indices] == min(fash_fit1_update$lfdr[selected_indices]))][1]
  fitted_result <- predict(fash_fit1_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  inv_weights <- (1/datasets[[selected_index]]$SE)/sum(1/datasets[[selected_index]]$SE)
  abline(h = sum(datasets[[selected_index]]$y*inv_weights), col = "blue", lty = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```

Even the most significant pairs from FASH for each gene do not show a clear pattern of dynamic eQTLs. 
It is kind of expected that FASH won't conclude with high confidence that these genes have dynamic eQTLs.



## Classifying dynamic eQTLs

Following the definition in Strober et.al, we will classify the detected dynamic eQTLs into different categories:

- *Early*: eQTLs with strongest effect during the first three days: $\max_{t\leq3} |\beta(t)| - \max_{t> 3} |\beta(t)| > 0$.

- *Late*: eQTLs with strongest effect during the last four days: $\max_{t\geq 12} |\beta(t)| - \max_{t< 12} |\beta(t)| > 0$.

- *Middle*: eQTLs with strongest effect during days 4-11: $\max_{4\leq t\leq 11} |\beta(t)| - \max_{t> 11 | t< 4} |\beta(t)| > 0$.

- *Switch*: eQTLs with effect sign switch during the time course such that $\min\{\max\beta(t)^+,\max\beta(t)^-\}-c>0$ where $c$ is a threshold that we set to 0.25 (which means with two alleles, the maximal difference of effect size is at least $\geq 2\times\min\{\max\beta(t)^+,\max\beta(t)^-\}\times2 \geq  2 \times 0.25  \times 2 = 1$).

We will take a look at the significant pairs detected by FASH (order 1), and classify them based on the false sign rate (lfsr).

### Early dynamic eQTLs

```{r eval=FALSE}
smooth_var_refined = seq(0,15, by = 0.1)
functional_early <- function(x){
  max(abs(x[smooth_var_refined <= 3])) - max(abs(x[smooth_var_refined > 3]))
}
testing_early_dyn <- testing_functional(functional_early,
                                              lfsr_cal = function(x){mean(x <= 0)},
                                              fash = fash_fit1,
                                              indices = fash_highlighted1,
                                              smooth_var = smooth_var_refined)
```

How many pairs and how many unique genes are classified as early dynamic eQTLs?

```{r}
load(paste0(result_dir, "/classify_dyn_eQTLs_early.RData"))
early_indices <- testing_early_dyn$indices[testing_early_dyn$cfsr <= alpha]
length(early_indices)
early_genes <- unique(sapply(strsplit(names(datasets)[early_indices], "_"), "[[", 1))
length(early_genes)
```

Let's take a look at the top-ranked early dynamic eQTLs:

```{r echo=FALSE}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- early_indices[i]
  fitted_result <- predict(fash_fit1,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  abline(h = 0, lty = 2, col = "blue")
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```

It is evident that these variants have largest $|\beta(t)|$ during the first three days.


### Middle dynamic eQTLs

```{r eval=FALSE}
functional_middle <- function(x){
  max(abs(x[smooth_var_refined <= 11 & smooth_var_refined >= 4])) - max(abs(x[smooth_var_refined > 11]), abs(x[smooth_var_refined < 4]))
}
testing_middle_dyn <- testing_functional(functional_middle, 
                                               lfsr_cal = function(x){mean(x <= 0)},
                                               fash = fash_fit1, 
                                               indices = fash_highlighted1, 
                                               num_cores = num_cores,
                                               smooth_var = smooth_var_refined)
```


How many pairs and how many unique genes are classified as middle dynamic eQTLs?
```{r}
load(paste0(result_dir, "/classify_dyn_eQTLs_middle.RData"))
middle_indices <- testing_middle_dyn$indices[testing_middle_dyn$cfsr <= alpha]
length(middle_indices)
middle_genes <- unique(sapply(strsplit(names(datasets)[middle_indices], "_"), "[[", 1))
length(middle_genes)
```

Take a look at their results:

```{r}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- sample(middle_indices, 1)
  fitted_result <- predict(fash_fit1,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  abline(h = 0, lty = 2, col = "blue")
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```


### Late dynamic eQTLs

```{r eval=FALSE}
functional_late <- function(x){
  max(abs(x[smooth_var_refined >= 12])) - max(abs(x[smooth_var_refined < 12]))
}
testing_late_dyn <- testing_functional(functional_late, 
                                             lfsr_cal = function(x){mean(x <= 0)},
                                             fash = fash_fit1, 
                                             indices = fash_highlighted1, 
                                             num_cores = num_cores,
                                             smooth_var = smooth_var_refined)
```

How many pairs and how many unique genes are classified as late dynamic eQTLs?

```{r}
load(paste0(result_dir, "/classify_dyn_eQTLs_late.RData"))
late_indices <- testing_late_dyn$indices[testing_late_dyn$cfsr <= alpha]
length(late_indices)
late_genes <- unique(sapply(strsplit(names(datasets)[late_indices], "_"), "[[", 1))
length(late_genes)
```

Let's take a look at the top-ranked late dynamic eQTLs:

```{r echo=FALSE}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- late_indices[i]
  fitted_result <- predict(fash_fit1,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  abline(h = 0, lty = 2, col = "blue")
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```


### Switch dynamic eQTLs

How many pairs and how many unique genes are classified as switch dynamic eQTLs?

```{r eval=FALSE}
switch_threshold <- 0.25
functional_switch <- function(x){
  # compute the radius of x, measured by deviation from 0 from below and from above
  x_pos <- x[x > 0]
  x_neg <- x[x < 0]
  if(length(x_pos) == 0 || length(x_neg) == 0){
    return(0)
  }
  min(max(abs(x_pos)), max(abs(x_neg))) - switch_threshold
}
testing_switch_dyn <- testing_functional(functional_switch, 
                                               lfsr_cal = function(x){mean(x <= 0)},
                                               fash = fash_fit1, 
                                               indices = fash_highlighted1, 
                                               num_cores = num_cores,
                                               smooth_var = smooth_var_refined)
```

```{r}
load(paste0(result_dir, "/classify_dyn_eQTLs_switch.RData"))
switch_indices <- testing_switch_dyn$indices[testing_switch_dyn$cfsr <= alpha]
length(switch_indices)
switch_genes <- unique(sapply(strsplit(names(datasets)[switch_indices], "_"), "[[", 1))
length(switch_genes)
```

Let's take a look at the top-ranked switch dynamic eQTLs:

```{r echo=FALSE}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- switch_indices[i]
  fitted_result <- predict(fash_fit1,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  abline(h = 0, lty = 2, col = "blue")
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
}
par(mfrow = c(1,1))
```







## Gene Set Enrichment Analysis

```{r warning=FALSE, message=FALSE}
library(clusterProfiler)
library(tidyverse)
library(msigdbr)
library(org.Hs.eg.db)  # Assuming human genes
library(biomaRt)
library(cowplot)
# Retrieve Hallmark gene sets for Homo sapiens
m_t2g <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

## A function to check gene-enrichment
enrich_set <- function(genes_selected, background_gene, q_val_cutoff = 0.05, pvalueCutoff = 0.05) {
  
  genes_converted <- getBM(
    filters = "ensembl_gene_id", 
    attributes = c("ensembl_gene_id", "entrezgene_id"), 
    values = genes_selected, 
    mart = mart
  )
  
  # Extract Entrez IDs from the converted data
  entrez_gene_list <- genes_converted$entrezgene_id
  
  genes_converted_all <- getBM(
    filters = "ensembl_gene_id", 
    attributes = c("ensembl_gene_id", "entrezgene_id"), 
    values = background_gene, 
    mart = mart
  )
  entrez_universe <- as.character(genes_converted_all$entrezgene_id)
  entrez_universe <- entrez_universe[!is.na(entrez_universe)]
  
  # Perform enrichment analysis using Hallmark gene sets
  enrich_res <- enricher(pAdjustMethod = "BH", 
                         entrez_gene_list, 
                         TERM2GENE = m_t2g, 
                         qvalueCutoff = q_val_cutoff, 
                         pvalueCutoff = pvalueCutoff, 
                         universe = entrez_universe)
  enrich_res
}
```

Among all the genes highlighted by FASH:

```{r}
result <- enrich_set(genes_selected = genes_highlighted1, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as early dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = early_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as middle dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = middle_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as late dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = late_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as switch dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = switch_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

