---
title: "Nonlinear Dynamic eQTL analysis on iPSC"
author: "Ziang Zhang"
date: "2025-02-21"
output: 
  workflowr::wflow_html:
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(fig.width = 8, fig.height = 6)
library(fashr)
library(biomaRt)
mart <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
result_dir <- paste0(getwd(), "/output/dynamic_eQTL_real")
data_dir <- paste0(getwd(), "/data/dynamic_eQTL_real")
code_dir <- paste0(getwd(), "/code/dynamic_eQTL_real")
log_prec <- seq(0,10, by = 0.2)
fine_grid <- sort(c(0, exp(-0.5*log_prec)))
```

## Obtain the effect size of eQTLs

We use the processed (expression \& genotype) data of [Strober et.al, 2019](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6623972/) to perform the eQTL analysis.

For the association testing, we use a linear regression model for each gene-variant pair at each time point.
Following the practice in Strober et.al, we adjust for the first five PCs.

The code to perform this step can be found in the script `dynamic_eQTL_real/00_eQTLs.R` from the code directory.

After this step, we have the effect size of eQTLs for each gene-variant pair at each time point, as well as its standard error.



## Fitting FASH

To fit the FASH model on $\{\beta_i(t_j), s_{ij}\}_{i\in N,j \in [16]}$, we consider fitting two FASH models:

- A FASH model based on first order IWP (testing for dynamic eQTLs: $H_0: \beta_i(t)=c$).

- A FASH model based on second order IWP (testing for nonlinear-dynamic eQTLs: $H_0: \beta_i(t)=c_1+c_2t$).

The code to perform this step can be found in the script `dynamic_eQTL_real/01_fash.R` from the code directory.

We will directly load the fitted FASH models from the output directory.

```{r}
load(paste0(result_dir, "/fash_fit2_all.RData"))
```

We will load the datasets from the fitted FASH object:

```{r}
datasets <- fash_fit2$fash_data$data_list
for (i in 1:length(datasets)) {
  datasets[[i]]$SE <- fash_fit2$fash_data$S[[i]]
}
all_genes <- unique(sapply(strsplit(names(datasets), "_"), "[[", 1))

full_map <- getBM(
  attributes = c("hgnc_symbol", "ensembl_gene_id"),
  filters    = "ensembl_gene_id",
  values     = all_genes,
  mart       = mart
)
```


In this analysis, we will focus on the **FASH(2)** model that assumes a **second order IWP** and tests for **nonlinear dynamic eQTLs**.

Let's take a quick overview of the fitted FASH model:

```{r, eval=FALSE}
log_prec <- seq(0,10, by = 0.2)
fine_grid <- sort(c(0, exp(-0.5*log_prec)))

fash_fit2 <- fash(Y = "beta", smooth_var = "time", S = "SE", data_list = datasets,
                  num_basis = 20, order = 2, betaprec = 0,
                  pred_step = 1, penalty = 10, grid = fine_grid,
                  num_cores = num_cores, verbose = TRUE)
save(fash_fit2, file = "./results/fash_fit2_all.RData")
```


```{r}
fash_fit2
```

As well as the estimated priors:
```{r}
fash_fit2$prior_weights
```




## Problem with $\pi_0$ estimation

The original MLE estimated $\pi_0$ is `r fash_fit2$prior_weights$prior_weight[1]`.
This could be under-estimated due to model-misspecification under the alternative hypothesis.
To account for this, we will consider the following approaches:

(i): A conservative estimate of $\pi_0$ based on the BF procedure:

```{r eval=FALSE}
fash_fit2_update <- BF_update(fash_fit2, plot = FALSE)
fash_fit2_update$prior_weights
save(fash_fit2_update, file = paste0(result_dir, "/fash_fit2_update.RData"))
```

```{r echo=FALSE}
load(paste0(result_dir, "/fash_fit2_update.RData"))
```

The conservative estimate is `r fash_fit2_update$prior_weights$prior_weight[1]`, which is much more conservative.

(ii): Instead of looking at the FDR which is based on the estimated $\pi_0$, we can use the minimum local false sign rate ($\text{min-lfsr}_i$) to measure significance:
$$
\text{min-lfsr}_i = \min_{t} \left\{ \text{lfsr}(W_i(t)) \right\},
$$
where $W_i(t) = \beta_i(t) - \beta_i(0)$.

## Detecting Nonlinear dynamic eQTLs

We will use the updated FASH model (2) to detect nonlinear dynamic eQTLs.

```{r}
alpha <- 0.05
test2 <- fdr_control(fash_fit2_update, alpha = alpha, plot = F)
fash_highlighted2 <- test2$fdr_results$index[test2$fdr_results$FDR <= alpha]

test2_before <- fdr_control(fash_fit2, alpha = alpha, plot = F)
fash_highlighted2_before <- test2_before$fdr_results$index[test2_before$fdr_results$FDR <= alpha]
```

How many pairs are detected?

```{r}
pairs_highlighted2 <- names(datasets)[fash_highlighted2]
length(pairs_highlighted2)
length(pairs_highlighted2)/length(datasets)
```


What is the number before the BF adjustment?

```{r}
pairs_highlighted2_before <- names(datasets)[fash_highlighted2_before]
length(pairs_highlighted2_before)
length(pairs_highlighted2_before)/length(datasets)
```


How many unique genes are detected?

```{r}
genes_highlighted2 <- unique(sapply(strsplit(pairs_highlighted2, "_"), "[[", 1))
length(genes_highlighted2)
length(genes_highlighted2)/length(all_genes)
```

Before the BF adjustment?

```{r}
genes_highlighted2_before <- unique(sapply(strsplit(pairs_highlighted2_before, "_"), "[[", 1))
length(genes_highlighted2_before)
length(genes_highlighted2_before)/length(all_genes)
```


Visualize top-ranked pairs:

```{r echo=FALSE}
genes_highlighted2_subset <- c("AMN", "RIPK2", "DENND5B", "ZIC2")
# convert to ensembl IDs
map <- getBM(
  attributes = c("hgnc_symbol", "ensembl_gene_id"),
  filters    = "hgnc_symbol",
  values     = genes_highlighted2_subset,
  mart       = mart
)
genes_highlighted2_subset_ens <- map$ensembl_gene_id[match(genes_highlighted2_subset, map$hgnc_symbol)]

par(mfrow = c(2,2))
for (i in 1:4) {
  selected_gene <- genes_highlighted2_subset_ens[i]  # sample(genes_highlighted2, 1)
  pairs_of_selected_gene <- grep(selected_gene, pairs_highlighted2, value = TRUE)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)

  selected_index <- selected_indices[
    which(fash_fit2_update$lfdr[selected_indices] ==
            min(fash_fit2_update$lfdr[selected_indices]))[1]
  ]

  # shorthand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)

  y_min <- min(y - 2 * se)
  y_max <- max(y + 2 * se)

  # weighted mean
  y_wmean <- sum(y * w) / sum(w)

  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)

  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))

  fitted_result <- predict(
    fash_fit2_update,
    index      = selected_index,
    smooth_var = seq(0, 15, by = 0.1)
  )

  plot(
    x, y,
    pch  = 20,
    col  = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = names(datasets)[selected_index],
    ylim = c(y_min, y_max)
  )
  
  # define the title to be the gene symbol: rs ID
  gene_symbol <- map$hgnc_symbol[map$ensembl_gene_id == selected_gene]
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  title(main = paste0(gene_symbol, ": ", variant_id))

  arrows(
    x0 = x,
    y0 = y - 2 * se,
    x1 = x,
    y1 = y + 2 * se,
    length = 0.05,
    angle  = 90,
    code   = 3,
    col    = "black"
  )

  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col    = rgb(1, 0, 0, 0.3),
    border = NA
  )

  # posterior mean curve
  lines(
    fitted_result$x,
    fitted_result$mean,
    col = "red",
    lwd = 2
  )

  # # weighted average line
  # abline(h = y_wmean, col = "blue",  lty = 2)

  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)

  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)
  
  # lfdr before and after
  lfdr_before <- fash_fit2$lfdr[selected_index]
  lfdr_after  <- fash_fit2_update$lfdr[selected_index]
  
  # text for caption
  caption <- sprintf("lfdr (raw) = %.3f, lfdr (adj) = %.3f",
                     lfdr_before, lfdr_after)
  
  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
 
}
par(mfrow = c(1,1))
```


Some examples of null pairs:

```{r}
genes_not_highlighted2 <- setdiff(all_genes, genes_highlighted2)
pairs_not_highlighted2 <- setdiff(names(datasets), pairs_highlighted2)
par(mfrow = c(2,2))
for (i in 10:13) {
  selected_gene <- genes_not_highlighted2[i]
  pairs_of_selected_gene <- grep(selected_gene, pairs_not_highlighted2, value = TRUE)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)

  # here we pick the *least* interesting one: max lfdr after adjustment
  selected_index <- selected_indices[
    which(fash_fit2_update$lfdr[selected_indices] ==
            max(fash_fit2_update$lfdr[selected_indices]))[1]
  ]

  # shorthand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)

  y_min <- min(y - 2 * se)
  y_max <- max(y + 2 * se)

  # weighted mean
  y_wmean <- sum(y * w) / sum(w)

  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)

  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))

  # FASH fit (after adjustment)
  fitted_result <- predict(
    fash_fit2_update,
    index      = selected_index,
    smooth_var = seq(0, 15, by = 0.1)
  )

  # lfdr before and after
  lfdr_before <- fash_fit2$lfdr[selected_index]
  lfdr_after  <- fash_fit2_update$lfdr[selected_index]

  # caption text
  caption <- sprintf("lfdr (raw) = %.3f, lfdr (adj) = %.3f",
                     lfdr_before, lfdr_after)

  plot(
    x, y,
    pch  = 20,
    col  = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = names(datasets)[selected_index],
    ylim = c(y_min, y_max)
  )
  
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  gene_symbol <- full_map$hgnc_symbol[full_map$ensembl_gene_id == selected_gene]
  title(main = paste0(gene_symbol, ": ", variant_id))

  arrows(
    x0 = x,
    y0 = y - 2 * se,
    x1 = x,
    y1 = y + 2 * se,
    length = 0.05,
    angle  = 90,
    code   = 3,
    col    = "black"
  )

  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col    = rgb(1, 0, 0, 0.3),
    border = NA
  )

  # posterior mean curve
  lines(
    fitted_result$x,
    fitted_result$mean,
    col = "red",
    lwd = 2
  )

  # # weighted average line (replaces simple mean)
  # abline(h = y_wmean, col = "blue",  lty = 2)

  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)

  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)

  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
}
par(mfrow = c(1,1))
```



## Comparing with Strober et.al

We will compare the detected dynamic eQTLs with the results from Strober et.al.

```{r echo=FALSE}
library(ggVennDiagram)
library(tidyverse)
strober_nonlinear <- read.delim(paste0(data_dir, "/strober_nonlinear/non_linear_dynamic_eqtls_5_pc.txt"))
strober_nonlinear$key <- paste0(strober_nonlinear$ensamble_id, "_", strober_nonlinear$rs_id)
strober_nonlinear_highlighted <- strober_nonlinear %>% filter(eFDR <= alpha)
pairs_highlighted_strober_nonlinear <- strober_nonlinear_highlighted$key
genes_highlighted_strober_nonlinear <- unique(sapply(strsplit(pairs_highlighted_strober_nonlinear, "_"), "[[", 1))
strober_linear <- read.delim(paste0(data_dir, "/strober_linear/linear_dynamic_eqtls_5_pc.txt"))
strober_linear$key <- paste0(strober_linear$ensamble_id, "_", strober_linear$rs_id)
strober_linear_highlighted <- strober_linear %>% filter(eFDR <= alpha)
pairs_highlighted_strober_linear <- strober_linear_highlighted$key
genes_highlighted_strober_linear <- unique(sapply(strsplit(pairs_highlighted_strober_linear, "_"), "[[", 1))
```

Let's take a look at the overlap between the two methods used in Strober et.al and FASH (order 2):

```{r}
gene_sets <- list(
  "Strober (Nonlinear)" = genes_highlighted_strober_nonlinear,
  "Strober (Linear)" = genes_highlighted_strober_linear,
  "FASH (2)" = genes_highlighted2
)
ggVennDiagram(gene_sets, label = "both") +
  scale_fill_gradient(low="grey90",high = "red") +
  theme(legend.position = "right")  # Move legend to the right
```

Let's take a look at the 4 pairs that are least significant from FASH:

```{r echo=FALSE}
gene_fashr_only <- setdiff(genes_highlighted2, union(genes_highlighted_strober_nonlinear, genes_highlighted_strober_linear))
## produce four plots for four genes
par(mfrow = c(2,2))
for (i in 1:min(c(4,length(gene_fashr_only)))) {
  selected_gene <- gene_fashr_only[i]
  pairs_of_selected_gene <- grep(selected_gene, pairs_highlighted2, value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit2_update$lfdr[selected_indices] == max(fash_fit2_update$lfdr[selected_indices]))][1]
  fitted_result <- predict(fash_fit2_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  gene_symbol <- full_map$hgnc_symbol[full_map$ensembl_gene_id == selected_gene]
  title(main = paste0(gene_symbol, ": ", variant_id))
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
  
  # lfdr before and after
  lfdr_before <- fash_fit2$lfdr[selected_index]
  lfdr_after  <- fash_fit2_update$lfdr[selected_index]
  
  # text for caption
  caption <- sprintf("lfdr (raw) = %.3f, lfdr (adj) = %.3f",
                     lfdr_before, lfdr_after)
  
  # short hand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)
  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)
  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))
  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)
  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)
  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
}
par(mfrow = c(1,1))
```

Let's also look at the genes that were missed by FASH, but detected by Strober et.al.
In this case, we will pick the most significant pair for each gene in FASH:

```{r echo=FALSE}
gene_missed <- setdiff(union(genes_highlighted_strober_nonlinear, genes_highlighted_strober_linear), genes_highlighted2)
## produce four plots for four genes
par(mfrow = c(2,2))
for (i in 1:min(c(4,length(gene_missed)))) {
  selected_gene <- gene_missed[i]
  pairs_of_selected_gene <- grep(selected_gene, names(datasets), value = T)
  selected_indices <- which(names(datasets) %in% pairs_of_selected_gene)
  selected_index <- selected_indices[which(fash_fit2_update$lfdr[selected_indices] == min(fash_fit2_update$lfdr[selected_indices]))][1]
  fitted_result <- predict(fash_fit2_update,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  gene_symbol <- full_map$hgnc_symbol[full_map$ensembl_gene_id == selected_gene]
  title(main = paste0(gene_symbol, ": ", variant_id))
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
  
  # lfdr before and after
  lfdr_before <- fash_fit2$lfdr[selected_index]
  lfdr_after  <- fash_fit2_update$lfdr[selected_index]
  # text for caption
  caption <- sprintf("lfdr (raw) = %.3f, lfdr (adj) = %.3f",
                     lfdr_before, lfdr_after)
  # short hand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)
  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)
  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))
  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)
  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)
  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
}
par(mfrow = c(1,1))
```


## Classifying nonlinear dynamic eQTLs

Following the definition in Strober et.al, we will classify the detected eQTLs into different categories:

- *Early*: eQTLs with strongest effect during the first three days: $\max_{t\leq3} |\beta(t)| - \max_{t> 3} |\beta(t)| > 0$.

- *Late*: eQTLs with strongest effect during the last four days: $\max_{t\geq 12} |\beta(t)| - \max_{t< 12} |\beta(t)| > 0$.

- *Middle*: eQTLs with strongest effect during days 4-11: $\max_{4\leq t\leq 11} |\beta(t)| - \max_{t> 11 | t< 4} |\beta(t)| > 0$.

- *Switch*: eQTLs with effect sign switch during the time course: $\min\{\max\beta(t)^+,\max\beta(t)^-\}-c $ where $c$ is a threshold that we set to 0.25 (which means with two alleles, the maximal difference of effect size is at least $\geq 2\times\min\{\max\beta(t)^+,\max\beta(t)^-\}\times2 \geq  2 \times 0.25  \times 2 = 1$).

We first take a look at the significant pairs detected by FASH (order 2), and classify them based on the false sign rate (lfsr):

```{r eval=FALSE}
smooth_var_refined = seq(0,15, by = 0.1)
functional_early <- function(x){
  max(abs(x[smooth_var_refined <= 3])) - max(abs(x[smooth_var_refined > 3]))
}
testing_early_nonlin_dyn <- testing_functional(functional_early,
                                              lfsr_cal = function(x){mean(x <= 0)},
                                              fash = fash_fit2,
                                              indices = fash_highlighted2,
                                              smooth_var = smooth_var_refined)
```

How many pairs and how many unique genes are classified as early dynamic eQTLs?

```{r}
load(paste0(result_dir, "/classify_nonlin_dyn_eQTLs_early.RData"))
early_indices <- testing_early_nonlin_dyn$indices[testing_early_nonlin_dyn$cfsr <= alpha]
length(early_indices)
early_genes <- unique(sapply(strsplit(names(datasets)[early_indices], "_"), "[[", 1))
length(early_genes)
```

How many pairs are classified as middle dynamic eQTLs?

```{r eval=FALSE}
functional_middle <- function(x){
  max(abs(x[smooth_var_refined <= 11 & smooth_var_refined >= 4])) - max(abs(x[smooth_var_refined > 11]), abs(x[smooth_var_refined < 4]))
}
testing_middle_nonlin_dyn <- testing_functional(functional_middle, 
                                               lfsr_cal = function(x){mean(x <= 0)},
                                               fash = fash_fit2, 
                                               indices = fash_highlighted2, 
                                               num_cores = num_cores,
                                               smooth_var = smooth_var_refined)
```


```{r}
load(paste0(result_dir, "/classify_nonlin_dyn_eQTLs_middle.RData"))
middle_indices <- testing_middle_nonlin_dyn$indices[testing_middle_nonlin_dyn$cfsr <= alpha]
length(middle_indices)
middle_genes <- unique(sapply(strsplit(names(datasets)[middle_indices], "_"), "[[", 1))
length(middle_genes)
```

Take a look at their results:

```{r}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- sample(middle_indices, 1)
  fitted_result <- predict(fash_fit2,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  gene_symbol <- full_map$hgnc_symbol[full_map$ensembl_gene_id == sapply(strsplit(names(datasets)[selected_index], "_"), "[[", 1)]
  title(main = paste0(gene_symbol, ": ", variant_id))
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
  
  # add lfsr
  lfsr_adj <- testing_middle_nonlin_dyn$lfsr[testing_middle_nonlin_dyn$indices == selected_index]
  
  caption <- sprintf("lfsr (adj) = %.3f",
                     lfsr_adj)

  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
  
  # shorthand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)
  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)
  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))
  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)
  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)
}
par(mfrow = c(1,1))
```


How many pairs are classified as late dynamic eQTLs?

```{r eval=FALSE}
functional_late <- function(x){
  max(abs(x[smooth_var_refined >= 12])) - max(abs(x[smooth_var_refined < 12]))
}
testing_late_nonlin_dyn <- testing_functional(functional_late, 
                                             lfsr_cal = function(x){mean(x <= 0)},
                                             fash = fash_fit2, 
                                             indices = fash_highlighted2, 
                                             num_cores = num_cores,
                                             smooth_var = smooth_var_refined)
```

```{r}
load(paste0(result_dir, "/classify_nonlin_dyn_eQTLs_late.RData"))
late_indices <- testing_late_nonlin_dyn$indices[testing_late_nonlin_dyn$cfsr <= alpha]
length(late_indices)
late_genes <- unique(sapply(strsplit(names(datasets)[late_indices], "_"), "[[", 1))
length(late_genes)
```

Let's take a look at the top-ranked late dynamic eQTLs:

```{r echo=FALSE}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- late_indices[i]
  fitted_result <- predict(fash_fit2,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  gene_symbol <- full_map$hgnc_symbol[full_map$ensembl_gene_id == sapply(strsplit(names(datasets)[selected_index], "_"), "[[", 1)]
  title(main = paste0(gene_symbol, ": ", variant_id))
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
  # shorthand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)
  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)
  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))
  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)
  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)
  
  # add lfsr
  lfsr_adj <- testing_late_nonlin_dyn$lfsr[testing_late_nonlin_dyn$indices == selected_index]
  caption <- sprintf("lfsr (adj) = %.3f",
                     lfsr_adj)
  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
}
par(mfrow = c(1,1))
```

How many pairs and how many unique genes are classified as switch dynamic eQTLs?

```{r eval=FALSE}
switch_threshold <- 0.25
functional_switch <- function(x){
  # compute the radius of x, measured by deviation from 0 from below and from above
  x_pos <- x[x > 0]
  x_neg <- x[x < 0]
  if(length(x_pos) == 0 || length(x_neg) == 0){
    return(0)
  }
  min(max(abs(x_pos)), max(abs(x_neg))) - switch_threshold
}
testing_switch_nonlin_dyn <- testing_functional(functional_switch, 
                                               lfsr_cal = function(x){mean(x <= 0)},
                                               fash = fash_fit2, 
                                               indices = fash_highlighted2, 
                                               num_cores = num_cores,
                                               smooth_var = smooth_var_refined)
```

```{r}
load(paste0(result_dir, "/classify_nonlin_dyn_eQTLs_switch.RData"))
switch_indices <- testing_switch_nonlin_dyn$indices[testing_switch_nonlin_dyn$cfsr <= alpha]
length(switch_indices)
switch_genes <- unique(sapply(strsplit(names(datasets)[switch_indices], "_"), "[[", 1))
length(switch_genes)
```

Let's take a look at the top-ranked switch dynamic eQTLs:

```{r echo=FALSE}
par(mfrow = c(2,2))
for (i in 1:4) {
  selected_index <- switch_indices[i]
  fitted_result <- predict(fash_fit2,
                           index = selected_index,
                           smooth_var = seq(0, 15, by = 0.1))
  plot(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y,
    pch = 20,
    col = "black",
    xlab = "Time",
    ylab = "Effect Est",
    # main = paste0(names(datasets)[selected_index]),
    ylim = c(
      min(datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE),
      max(datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE)
    )
  )
  variant_id <- strsplit(names(datasets)[selected_index], "_")[[1]][2]
  gene_symbol <- full_map$hgnc_symbol[full_map$ensembl_gene_id == sapply(strsplit(names(datasets)[selected_index], "_"), "[[", 1)]
  title(main = paste0(gene_symbol, ": ", variant_id))
  arrows(
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y - 2 * datasets[[selected_index]]$SE,
    datasets[[selected_index]]$x,
    datasets[[selected_index]]$y + 2 * datasets[[selected_index]]$SE,
    length = 0.05,
    angle = 90,
    code = 3,
    col = "black"
  )
  lines(fitted_result$x,
        fitted_result$mean,
        col = "red",
        lwd = 2)
  polygon(
    c(fitted_result$x, rev(fitted_result$x)),
    c(fitted_result$lower, rev(fitted_result$upper)),
    col = rgb(1, 0, 0, 0.3),
    border = NA
  )
  # shorthand
  x  <- datasets[[selected_index]]$x
  y  <- datasets[[selected_index]]$y
  se <- datasets[[selected_index]]$SE
  w  <- 1 / (se^2)
  # weighted linear fit
  lin_fit <- lm(y ~ x, weights = w)
  # weighted quadratic fit
  quad_fit  <- lm(y ~ poly(x, 2, raw = TRUE), weights = w)
  x_grid    <- seq(min(x), max(x), length.out = 200)
  quad_pred <- predict(quad_fit, newdata = data.frame(x = x_grid))
  # weighted linear fit
  abline(lin_fit,     col = "green", lty = 2, lwd = 1)
  # weighted quadratic curve
  lines(x_grid, quad_pred, col = "purple", lty = 2, lwd = 1)
  # add lfsr
  lfsr_adj <- testing_switch_nonlin_dyn$lfsr[testing_switch_nonlin_dyn$indices == selected_index]
  caption <- sprintf("lfsr (adj) = %.3f",
                     lfsr_adj)
  # add caption as subtitle
  title(sub = caption, cex.sub = 0.8)
}
par(mfrow = c(1,1))
```



## Gene Set Enrichment Analysis

```{r warning=FALSE, message=FALSE}
library(clusterProfiler)
library(tidyverse)
library(msigdbr)
library(org.Hs.eg.db)  # Assuming human genes
library(biomaRt)
library(cowplot)
# Retrieve Hallmark gene sets for Homo sapiens
m_t2g <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

## A function to check gene-enrichment
## Retrieve Hallmark gene sets for Homo sapiens (use Ensembl IDs)
m_t2g <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  mutate(ensembl_use = dplyr::coalesce(ensembl_gene, db_ensembl_gene)) %>% 
  dplyr::filter(!is.na(ensembl_use)) %>% 
  dplyr::select(gs_name, ensembl_use) %>%
  dplyr::distinct()    # one row per (pathway, Ensembl) pair


## A function to check gene-enrichment using Ensembl IDs,
## forcing the universe to be exactly background_gene
enrich_set <- function(genes_selected,
                       background_gene,
                       q_val_cutoff = 0.05,
                       pvalueCutoff = 0.05) {

  # ensure character vectors & unique
  genes_selected_raw  <- unique(as.character(genes_selected))
  background_gene_raw <- unique(as.character(background_gene))

  # we'll enforce that the test universe is exactly these:
  universe_for_test <- background_gene_raw

  # genes that are in Hallmark already
  hallmark_genes <- unique(m_t2g$ensembl_use)

  # background genes that are NOT in Hallmark (no annotation)
  bg_not_in_hallmark <- setdiff(universe_for_test, hallmark_genes)

  # extend TERM2GENE so that *all* background genes appear in it at least once
  # add them to a dummy pathway that we will later drop
  dummy_id <- "__DUMMY_BACKGROUND__"

  if (length(bg_not_in_hallmark) > 0) {
    dummy_t2g <- tibble(
      gs_name     = dummy_id,
      ensembl_use = bg_not_in_hallmark
    )
    TERM2GENE_full <- bind_rows(m_t2g, dummy_t2g)
  } else {
    TERM2GENE_full <- m_t2g
  }

  # also make sure selected genes are a subset of the universe
  genes_sel_used <- intersect(genes_selected_raw, universe_for_test)

  # message("Original selected genes: ", length(genes_selected_raw),
  #         " ; used in enrichment: ", length(genes_sel_used))
  # message("Universe size (background_gene): ", length(universe_for_test))

  enrich_res <- enricher(
    gene          = genes_sel_used,
    TERM2GENE     = TERM2GENE_full,
    universe      = universe_for_test,
    pAdjustMethod = "BH",
    qvalueCutoff  = q_val_cutoff,
    pvalueCutoff  = pvalueCutoff
  )

  if (is.null(enrich_res) || nrow(enrich_res@result) == 0L) {
    return(enrich_res)
  }

  df <- enrich_res@result

  # drop the dummy term from results
  df <- df %>% dplyr::filter(ID != dummy_id)

  # keep original ratios for reference
  df$GeneRatio_orig <- df$GeneRatio
  df$BgRatio_orig   <- df$BgRatio

  # now recompute ratios using exactly:
  #   - denominator for GeneRatio = number of selected genes in universe
  #   - denominator for BgRatio   = number of background genes
  n_sel_total <- length(genes_sel_used)
  n_bg_total  <- length(universe_for_test)

  df$GeneRatio_fixed <- paste0(df$Count,   "/", n_sel_total)
  df$BgRatio_fixed   <- paste0(df$setSize, "/", n_bg_total)

  enrich_res@result <- df
  enrich_res
}
```

Among all the genes highlighted by FASH:

```{r}
result <- enrich_set(genes_selected = genes_highlighted2, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as middle dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = middle_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as late dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = late_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

Among the genes highlighted by FASH that are classified as switch dynamic eQTLs:

```{r}
result <- enrich_set(genes_selected = switch_genes, background_gene = all_genes)
result@result %>% 
  filter(pvalue < 0.05) %>%
  dplyr::select(GeneRatio, BgRatio, pvalue, qvalue)
```

